src/server/endpoints/covid_hosp_state_timeseries.py

```diff
@@ -152,19 +152,20 @@ def handle():
     q.where_integers("date", dates)
     q.where_strings("state", states)
 
+    merge_tables = "(SELECT *, 'D' as record_type FROM `covid_hosp_state_daily` UNION ALL SELECT *, 'T' as record_type FROM `covid_hosp_state_timeseries`) c"
```

Comment:
the logic is perfect, but i fear that using this approach will cause a materialization of the entirety of the underlying tables...  can you do some EXPLAINs on the resulting queries to see if thats the case?

Reply:
```
mysql> EXPLAIN SELECT *, 'D' as record_type FROM `covid_hosp_state_daily` UNION ALL SELECT *, 'T' as record_type FROM `covid_hosp_state_timeseries`;
+----+-------------+-----------------------------+------------+------+---------------+------+---------+------+------+----------+-------+
| id | select_type | table                       | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra |
+----+-------------+-----------------------------+------------+------+---------------+------+---------+------+------+----------+-------+
|  1 | PRIMARY     | covid_hosp_state_daily      | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    3 |   100.00 | NULL  |
|  2 | UNION       | covid_hosp_state_timeseries | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | NULL  |
+----+-------------+-----------------------------+------------+------+---------------+------+---------+------+------+----------+-------+
2 rows in set, 1 warning (0.01 sec)
```
```
mysql> EXPLAIN ANALYZE SELECT *, 'D' as record_type FROM `covid_hosp_state_daily` UNION ALL SELECT *, 'T' as record_type FROM `covid_hosp_state_timeseries`;
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Append  (actual time=0.086..0.153 rows=4 loops=1)
    -> Stream results  (cost=0.55 rows=3) (actual time=0.086..0.100 rows=3 loops=1)
        -> Table scan on covid_hosp_state_daily  (cost=0.55 rows=3) (actual time=0.071..0.078 rows=3 loops=1)
    -> Stream results  (cost=0.35 rows=1) (actual time=0.049..0.050 rows=1 loops=1)
        -> Table scan on covid_hosp_state_timeseries  (cost=0.35 rows=1) (actual time=0.044..0.045 rows=1 loops=1)
 |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
Doesn't look like there's a materialization step here.

Reply:
Here's a query plan for that:
```
mysql> EXPLAIN ANALYZE WITH c AS (SELECT * , ROW_NUMBER() OVER (PARTITION BY date, state, issue ORDER BY record_type) `row` FROM (SELECT *, 'D' as record_type FROM `covid_hosp_state_daily` UNION ALL SELECT *, 'T' as record_type FROM `covid_hosp_state_timeseries`) c) SELECT * FROM c;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Table scan on c  (cost=2.50..2.50 rows=0) (actual time=0.233..0.238 rows=4 loops=1)
    -> Materialize CTE c  (cost=0.00..0.00 rows=0) (actual time=0.231..0.231 rows=4 loops=1)
        -> Window aggregate: row_number() OVER (PARTITION BY c.`date`,c.state,c.issue ORDER BY c.record_type )   (actual time=0.185..0.199 rows=4 loops=1)
            -> Sort: c.`date`, c.state, c.issue, c.record_type  (cost=5.05..5.05 rows=4) (actual time=0.176..0.179 rows=4 loops=1)
                -> Table scan on c  (cost=1.94..3.85 rows=4) (actual time=0.154..0.159 rows=4 loops=1)
                    -> Union all materialize  (cost=1.30..1.30 rows=4) (actual time=0.152..0.152 rows=4 loops=1)
                        -> Table scan on covid_hosp_state_daily  (cost=0.55 rows=3) (actual time=0.069..0.076 rows=3 loops=1)
                        -> Table scan on covid_hosp_state_timeseries  (cost=0.35 rows=1) (actual time=0.044..0.045 rows=1 loops=1)
 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```
And an alternate version of that query which puts the `UNION ALL` in the `WITH` clause:
```
mysql> EXPLAIN ANALYZE WITH merge_tables AS (SELECT *, 'D' as record_type FROM `covid_hosp_state_daily` UNION ALL SELECT *, 'T' as record_type FROM `covid_hosp_state_timeseries`), c AS (SELECT * , ROW_NUMBER() OVER (PARTITION BY date, state, issue ORDER BY record_type) `row` FROM merge_tables) SELECT * FROM c;
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| EXPLAIN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Table scan on c  (cost=2.50..2.50 rows=0) (actual time=0.232..0.237 rows=4 loops=1)
    -> Materialize CTE c  (cost=0.00..0.00 rows=0) (actual time=0.230..0.230 rows=4 loops=1)
        -> Window aggregate: row_number() OVER (PARTITION BY merge_tables.`date`,merge_tables.state,merge_tables.issue ORDER BY merge_tables.record_type )   (actual time=0.183..0.198 rows=4 loops=1)
            -> Sort: merge_tables.`date`, merge_tables.state, merge_tables.issue, merge_tables.record_type  (cost=5.05..5.05 rows=4) (actual time=0.176..0.178 rows=4 loops=1)
                -> Table scan on merge_tables  (cost=1.94..3.85 rows=4) (actual time=0.155..0.160 rows=4 loops=1)
                    -> Materialize union CTE merge_tables  (cost=1.30..1.30 rows=4) (actual time=0.153..0.153 rows=4 loops=1)
                        -> Table scan on covid_hosp_state_daily  (cost=0.55 rows=3) (actual time=0.069..0.077 rows=3 loops=1)
                        -> Table scan on covid_hosp_state_timeseries  (cost=0.35 rows=1) (actual time=0.044..0.045 rows=1 loops=1)
 |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Reply:
yuck, that looks like its gonna be pretty bad.  plus that EXPLAIN is for just one of three query forms that are used. 
 this is going to need some work, like distributing/factoring conditions across or into subqueries.  honestly, the existing query forms seem like they are more complicated than they need to be, so perhaps rewriting them is the way to go.